### PR TITLE
Update batch norm freezing to freeze running stats

### PR DESCRIPTION
**Pull request by joelnkn**
This PR seeks to update all areas in chemprop where batch norm layers were previously frozen using gradients alone. This does not freeze the "running stats" (running mean and variance) kept track of in the module, as discussed in #904. This can be fixed by setting the batch norm module to evaluation mode, ie. add `model.bn.eval()`.

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Review by KnathanM — COMMENTED** [review]
I see that `fingerprint.py` has a line `model.eval()` and `mpnn_fingerprint.ipynb` has `mpnn.eval()`. Is that enough to set the batch norm layer to eval, or do we also need the `bn.eval()` call that you added. 

I also wonder if `model.bn.apply(lambda module: module.requires_grad_(False))` in `train.py` isn't needed. Does setting `bn.eval()` also disable the gradients for the batch norm layer? If it does, a follow up question is if we need the lines
```
model.message_passing.apply(lambda module: module.requires_grad_(False))
model.message_passing.apply(
    lambda m: setattr(m, "p", 0.0) if isinstance(m, torch.nn.Dropout) else None
)
```
or if we can just do `model.message_passing.eval()`. 

And if it works for message passing, can we also do something like `model.predictor.ffn[idx].eval()` instead of 
```
model.predictor.ffn[idx].requires_grad_(False)
setattr(model.predictor.ffn[idx + 1][1], "p", 0.0)
```

We used `requires_grad_(False)` and `setattr(..., "p", 0.0)` because I thought that lightning set the whole model to training model at the beginning of `trainer.fit`, but lightning keeps track of which parts of the model are in eval mode and honors that setting, then we could simplify the transfer learning code.

**Comment by joelnkn** [comment]
Thanks for the comment @KnathanM. 

For the first point, `model.eval()` alone is sufficient, I missed those lines earlier. 

For the second point, after reading a few discussion boards, it looks like `model.eval()` alone, without gradients disabled, will not necessarily stop gradients from being calculated. It is mainly for changing the `forward()` behavior regarding Dropout and BatchNorm layers. Disabling gradients ends up saving a significant portion of memory. This does mean we can remove the lines setting `p` in Dropout layers as this is redundant in `eval()` mode. 

Related discussion [here](https://discuss.pytorch.org/t/should-i-use-model-eval-when-i-freeze-batchnorm-layers-to-finetune/39495/3).

**Comment by KnathanM** [comment]
That makes sense, so it looks like for transfer learning we will for sure want to use `requires_grad_(False)` for the frozen layers because the optimizer will be called and that would update the model weights, even if they are set to eval. For fingerprinting, we can continue to use `with torch.no_grad():` though we could also use `with torch.inference_mode()`. 

I think then the last thing I'd like to see before merging this is that the weights and buffers of frozen layers (message passing, batch normalization, any frozen ffn layers) do actually not change when using this transfer learning code. Could you test this using a tiny model (hidden size something like 4) and post your code here with the results?

**Comment by joelnkn** [comment]
https://pastebin.com/71fdKZgH

Here's a Pastebin with a Jupyter notebook that shows all the model buffers and parameters before and after training with the new layer freezing procedure. It shows everything freezing correctly up to some specified layer.

**Review by KnathanM — APPROVED** [review]
That looks great Joel, thanks for putting that together. If you are ready for me to merge this, mark it as ready. Also can you make the same changes (of .eval() instead of setattr(m, "p"...)  in #904? Then I can merge that one as well.

**Review by davidegraff — COMMENTED** [review]
So some questions I have on the changes in this PR:
1. If we're using a "frozen" model where some `Module`s are set to `train()` and others to `eval()`, does Lightning respect these modes in `Trainer.fit()`? Because calling `MPNN.train()` will recursively set all modules to `train()`, overriding the modes we've manually set
2. Is there a reason we (previously) set the dropout probability to `0` in the case of frozen models?
3. Is there a reason we desire to have the expected mean/variance statistics frozen now? Should these not adapt to the new training distribution? Presumably, there will be minor distribution shift between "pretraining" and "training," so why should we use the "old" statistics?

**Comment by KnathanM** [comment]
1. Apparently Lightning does respect it. In the `training.ipynb` notebook, I set `mpnn.message_passing.eval()`. I also added `print(self.message_passing.training)` to `training_step` in `MPNN`. During `trainer.fit(...)`, `False` gets printed on every step.
2. We previously set the dropout to 0 because I thought Lightning didn't respect having some modules in eval (because of the behavior of `MPNN.train()` as you pointed out). 
3. A user can choose whether to freeze the encoding (output of mp/agg/bn) or not. If they think their input distribution for fine-tuning is about the same as pretraining, then it is reasonable to freeze the encoder and always get the same encoding for a molecule. They can even freeze some of the ffn. If the input distribution is pretty different, then they can load the pretrained weights but not freeze any of them. 

This [stack overflow](https://stackoverflow.com/questions/70998847/transfer-learning-fine-tuning-how-to-keep-batchnormalization-in-inference-mode) quotes a tensorflow tutorial that says batchnorm should be set to inference mode during fine tuning.

**Comment by davidegraff** [comment]
Thanks for checking (1) and (2). This new approach is much cleaner, so it's a nice change. re: (3), this might be down to how we logically define "encoding" in chemprop. I think I might've defined the `encode()` function as $\verb|batch_norm|\circ \mathtt{agg} \circ \verb|message_passing|$, but IMO we shouldn't think of $\verb|batch_norm|$ as part of the "encoding" of an MPNN. It's a regularizer that normalizes the inputs to the FFN to stabilize training. That is:

$$
\mathbf z = \mathtt{agg} \circ \verb|message_passing|(\mathcal G)
$$

$$
\mathbf z_\text{norm} = \verb|batch_norm|(\mathbf z)
$$

where $\mathbf z$ is the "encoding" or "latent representation" of the input and $\mathbf z_\text{norm}$ is simply some regularized version of $\mathbf z$ used as input to the FFN.

I have two opinions: (1) choice of freezing one or both of the "encoder" and the batch norm should be decoupled because (2) the choice of freezing batch norm is empirical. The TF tutorial mentioned in your SO post includes no references, and whether or not to freeze BN is an inductive bias:

- freeze: when you _don't_ expect covariate shift, i.e., your pretraining and training **input** distributions _should_ be identical. This is **not** related to whether you think the input-output mapping (i.e., label shift) is the same.
- don't freeze: when you expect distribution shift. BN works on the premise of unit-normalizing each feature across a batch of inputs into the succeeding layer. If you reasonably expect covariate shift (which BTW is _fine_ for transfer learning or fine-tuning), then freezing your BN statistics should _harm_ training.

But again, this is ML at the end of the day, and our theoretical expectations can only take us so far. These models are incredibly, and even when we hamstring them, they typically find a way around that in ways that we _can't_ notice (until it's too late...)

**Comment by KnathanM** [comment]
I agree with you that $\mathbf{z}$ (before batch norm) is the encoding while after batch norm is some regularized version of the encoding. We can add a flag for whether to freeze the batch norm or not in the CLI and can show that it is optional in the example notebook. Indeed [this](https://forums.fast.ai/t/should-batch-norm-statistics-be-updated-during-transfer-learning/78800) post online references a paper where they found that resetting the batch norm (gamma & sigma = 1, mu & beta = 0) and training it new worked best for their transfer learning, which I think is similar to reinitializing some layers of the FFN instead of starting from the weights of the previous model. 

I think should also enforce in the CLI that you have to freeze the batch norm if any of the FFN layers are frozen. This isn't strictly true, but I think this is more likely a user mistake than intentional.

Outside the scope of transfer learning, I am unsure how to best give users the option to get the encoding before batch norm. Right now the `encoding()` method of `MPNN` uses `.fingerprint()` which includes batch normalization. Maybe this is okay to not give a function to get the pre-regularized encodings because often these encodings are used for clustering which I don't think is affected by scaling and shifting the encodings. Also the users can do something similar to what we do in the `mpnn_fingerprints.ipynb` example if they want the pre-regularized encodings.
```
with torch.no_grad():
    H = [
        mpnn.agg(mpnn.message_passing(batch.bmg, batch.V_d), batch.bmg.batch)
        for batch in test_loader
    ]
    encodings = torch.cat(H, 0)
```

-------------
Side note for future reference, I think this is the lightning class that allows it to restore the training mode of submodules during training. https://github.com/Lightning-AI/pytorch-lightning/blob/29f70e7d474b17c09c0c7343168bab5708335bd2/src/lightning/pytorch/utilities/model_helpers.py#L64
I found this as I was trying to reconcile the changes required in #822. It looks like during validation, lightning sets the whole model to evaluation mode, but during training it only sets to train those modules that were previously in training mode. This makes sense to me as you wouldn't have part of your model in training mode to test its performance but you could have parts of your model frozen during training for transfer learning. Except we do want our output_transform class to be in train mode during validation so we use the hook lightning provides.

**Comment by KnathanM** [comment]
I agree that putting the batch normalization after concatenating the MPNN output and the extra molecule descriptors into a fingerprint makes sense. At the same time, it may not matter too much if the extra molecule descriptors are properly normalized before training.

I've also see people in CNNs use multiple batch normalization layers. There is also layer normalization. We also found that some of the benchmarks for the v2 paper did surprisingly worse with batch normalization. All this is to say, I think we could do a whole overhaul update PR of normalization between layers in the model, but that would require doing some tests to see what empirically helps and what doesn't on a broad range of test sets. My preference would be to move this PR forward and save those batch normalization changes for another day.
